lib/internal/webstreams/readablestream.js

```diff
@@ -1249,6 +1256,75 @@ const isReadableStreamBYOBReader =
 
 // ---- ReadableStream Implementation
 
+function readableStreamFromIterable(iterable) {
+  let stream;
+  const iteratorRecord = getIterator(iterable, 'async');
+
+  const startAlgorithm = nonOpStart;
+
+  function pullAlgorithm() {
+    let nextResult;
+    try {
+      nextResult = iteratorNext(iteratorRecord);
+    } catch (error) {
+      return PromiseReject(error);
+    }
+    const nextPromise = PromiseResolve(nextResult);
+    return PromisePrototypeThen(nextPromise, (iterResult) => {
+      if (typeof iterResult !== 'object' || iterResult === null) {
+        throw new ERR_INVALID_STATE.TypeError(
+          'The promise returned by the iterator.next() method must fulfill with an object');
+      }
+      const done = Boolean(iterResult.done);
+      if (done) {
+        readableStreamDefaultControllerClose(stream[kState].controller);
+      } else {
+        readableStreamDefaultControllerEnqueue(stream[kState].controller, iterResult.value);
+      }
+    });
+  }
+
+  function cancelAlgorithm(reason) {
+    const iterator = iteratorRecord.iterator;
+    let returnMethod;
+    try {
+      returnMethod = iterator.return;
```

Comment:
Can't this be an async function which would allow the catch to not exist nd the below return PromiseResolve() to just be a return etc?

Reply:
Hmm i think it could, let me make it so and see if the WPTs pass

Reply:
Hey! updated this, the WPTs don't complain and reading the spec i doesn't seem to say anything about the function having to sync or async

Reply:
Indeed, this is equivalent. I wanted the error handling to be very explicit [in the reference implementation](https://github.com/whatwg/streams/blob/8d7a0bf26eb2cc23e884ddbaac7c1da4b91cf2bc/reference-implementation/lib/abstract-ops/readable-streams.js#L1914-L1928), but practical implementations can be more succinct. 🙂